src/budget/budgetmanager.cpp

```diff
@@ -1087,52 +1102,59 @@ int CBudgetManager::ProcessFinalizedBudgetVote(CFinalizedBudgetVote& vote, CNode
-        std::string strError = "masternode not valid or PoSe banned";
-        if (mnList.IsMNValid(dmn) && UpdateFinalizedBudget(vote, pfrom, strError)) {
-            vote.Relay();
-            const uint256& voteID = vote.GetHash();
-            masternodeSync.AddedBudgetItem(voteID);
-            LogPrint(BCLog::MNBUDGET, "fbvote - new vote (%s) for budget %s from dmn %s\n",
-                    voteID.ToString(), vote.GetBudgetHash().ToString(), dmn->proTxHash.ToString());
-        } else {
-            LogPrint(BCLog::MNBUDGET, "fbvote - error: %s", strError);
+        strError = "masternode not valid or PoSe banned";
+        if (!mnList.IsMNValid(dmn) || !UpdateFinalizedBudget(vote, pfrom, strError)) {
+            LogPrint(BCLog::MNBUDGET, "fbvote - error: %s\n", strError);
+            return false;
         }
-        return 0;
+
+        vote.Relay();
+        masternodeSync.AddedBudgetItem(voteID);
+        LogPrint(BCLog::MNBUDGET, "fbvote - new vote (%s) for budget %s from dmn %s\n",
+                voteID.ToString(), vote.GetBudgetHash().ToString(), dmn->proTxHash.ToString());
+        return true;
     }
 
     // -- Legacy System (!TODO: remove after enforcement) --
     CMasternode* pmn = mnodeman.Find(voteVin.prevout);
     if (!pmn) {
-        LogPrint(BCLog::MNBUDGET, "fbvote - unknown masternode - vin: %s\n", voteVin.prevout.hash.ToString());
+        strError = strprintf("unknown masternode - vin: %s", voteVin.prevout.ToString());
+        LogPrint(BCLog::MNBUDGET, "fbvote - error: %s\n", strError);
         mnodeman.AskForMN(pfrom, voteVin);
-        return 0;
+        return false;
     }
 
     AddSeenFinalizedBudgetVote(vote);
 
     if (!vote.CheckSignature(pmn->pubKeyMasternode.GetID())) {
         if (masternodeSync.IsSynced()) {
-            LogPrint(BCLog::MNBUDGET, "fbvote - signature from masternode %s invalid\n", HexStr(pmn->pubKeyMasternode));
-            return 20;
+            strError = strprintf("signature from masternode %s invalid", voteVin.prevout.ToString());
+            LogPrint(BCLog::MNBUDGET, "fbvote - error: %s\n", strError);
+            nDos = 20;
+        } else {
+            // it could just be a non-synced masternode
+            mnodeman.AskForMN(pfrom, voteVin);
         }
-        // it could just be a non-synced masternode
-        mnodeman.AskForMN(pfrom, voteVin);
-        return 0;
+        return false;
     }
 
-    std::string strError;
-    if (UpdateFinalizedBudget(vote, pfrom, strError)) {
-        vote.Relay();
-        masternodeSync.AddedBudgetItem(vote.GetHash());
-        LogPrint(BCLog::MNBUDGET, "fbvote - new finalized budget vote - %s from masternode %s\n", vote.GetHash().ToString(), HexStr(pmn->pubKeyMasternode));
-    } else {
-        LogPrint(BCLog::MNBUDGET, "fbvote - rejected finalized budget vote - %s from masternode %s - %s\n", vote.GetHash().ToString(), HexStr(pmn->pubKeyMasternode), strError);
+    strError = "masternode not valid";
```

Comment:
same as above, makes no difference (and you couldn't change it inside the `if`, because it could have been set by `UpdateProposal`/`UpdateFinalizedBudget` at that point). 
Will refactor in a clearer way.

Reply:
k, thanks